### PR TITLE
Add WithState to Mock Request Builder

### PR DESCRIPTION
**Pull request by henryou007**

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -632,10 +632,10 @@ func (a *App) onResponse(writer *logger.Writer, ts logger.TimeSource, eventFlag
 // renderAction is the translation step from Action to Handler.
 // this is where the bulk of the "pipeline" happens.
 func (a *App) renderAction(action Action) Handler {
-	return func(w http.ResponseWriter, r *http.Request, route *Route, p RouteParameters, tx *sql.Tx) {
+	return func(w http.ResponseWriter, r *http.Request, route *Route, p RouteParameters, tx *sql.Tx, state State) {
```
Comment (wcharczuk) [inline app.go]: get rid of this tx; just extract it from state later

File: app.go [hunk 2]
```diff
@@ -671,16 +671,16 @@ func (a *App) shouldCompressOutput(r *http.Request) bool {
...
 	context.onRequestStart()
 	a.logger.OnEvent(logger.EventWebRequestStart, context)
 	return context
 }
 
 // Ctx creates a context.
-func (a *App) newCtx(w ResponseWriter, r *http.Request, route *Route, p RouteParameters) *Ctx {
-	ctx := NewCtx(w, r, p)
+func (a *App) newCtx(w ResponseWriter, r *http.Request, route *Route, p RouteParameters, s State) *Ctx {
+	ctx := NewCtx(w, r, p, s) //a.ctxPool.Get()
```
Comment (wcharczuk) [inline app.go]: can get rid of this commented section

File: mock_request_builder.go [hunk 3]
```diff
@@ -52,6 +53,7 @@ type MockRequestBuilder struct {
 
 	responseBuffer *bytes.Buffer
 	tx             *sql.Tx
```
Comment (wcharczuk) [inline mock_request_builder.go]: ditch this tx, it should just be set in the map

**Comment by wcharczuk** [comment]
overall; i would ditch the `tx *sql.Tx` references that liter the app and ctx types. i would add a `StateProvider` interface like we spec'd in the playground, i would also add a helper to extract the Tx from the state.

**Comment by maxponte** [comment]
Is this so that you can pass { test: true } etc. to the handler

File: ctx.go [hunk 4]
```diff
@@ -67,19 +67,35 @@ type Ctx struct {
...
-
-	tx *sql.Tx
 }
 
-// WithTx sets a transaction on the context.
-func (rc *Ctx) WithTx(tx *sql.Tx) *Ctx {
-	rc.tx = tx
-	return rc
+// DBTransactionStateProvider stores the db transactions for Ctx
+type DBTransactionStateProvider interface {
```
Comment (wcharczuk) [inline ctx.go]: This can just be `StateProvider`

File: ctx.go [hunk 5]
```diff
@@ -67,19 +67,35 @@ type Ctx struct {
...
+	State(string) interface{}
+	SetState(key string, value interface{})
 }
 
 // Tx returns the transaction for the request.
-func (rc *Ctx) Tx() *sql.Tx {
-	return rc.tx
+func Tx(sp DBTransactionStateProvider, keys ...string) *sql.Tx {
```
Comment (wcharczuk) [inline ctx.go]: can you move both of these into a `state.go` file or similar

File: state.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import "database/sql"
+
+// StateProvider provide states, an example is Ctx
+type StateProvider interface {
+	State(string) interface{}
+	SetState(key string, value interface{})
+}
+
+// Tx returns the transaction for the request.
+func Tx(sp StateProvider, keys ...string) *sql.Tx {
```
Comment (wcharczuk) [inline state.go]: i would add one more method `WithTx()` that sets the tx in the state provider based on defaults

**Pull request merged**